compiler/parser/RdrHsSyn.hs

```diff
@@ -3092,75 +3112,74 @@ mkTemplateDecls templateName fields decls = do
-      templateImplementsMarkers = concatMap (mkImplementsMarker vtTemplateName) vtImplements
+      adjustImplementsDeclBlock implements =
+        let newDefs =
+              [ L loc (ImplementsFunction (name, newExp))
+              | L loc (ImplementsFunction (name, exp)) <- implementsDefs implements
+              , let this = asPatRecWild "this" conName
+                    args = [this]
+                    newExp = mkLambda args exp (allLetBindings True args sharedBinds)
+              ]
+        in implements { implementsDefs = newDefs }
+      templateInterfaceImplements =  (mkInterfaceImplements vtTemplateName . adjustImplementsDeclBlock) =<< vtImplements
   return $ toOL $ templateDataDecl : choiceDataDecls ++ letDecls
                ++ templateInstances ++ (choiceInstanceDecls ++ choiceDecls ++ keyInstanceDecl)
-               ++ templateInterfaceImplements ++ templateImplementsMarkers
-
-mkImplementsMarker :: Located RdrName -> ImplementsDeclBlock -> [LHsDecl GhcPs]
-mkImplementsMarker template ImplementsDeclBlock{implementsInterface=interface} =
-    [ noLoc $ SigD noExt (TypeSig noExt [name] ty)
-    , noLoc $ ValD noExt $ FunBind
-        noExt name
-        (matchGroup noSrcSpan $
-         matchWithBinds (matchContext name) [] noSrcSpan (mkQualVar (mkDataOcc "Implements")) (noLoc emptyLocalBinds))
-        WpHole []
-    ]
+               ++ templateInterfaceImplements
+
+mkInterfaceImplements :: Located RdrName -> ImplementsDeclBlock -> [LHsDecl GhcPs]
+mkInterfaceImplements templateName implements =
+  implementsInstance : implementsInterfaceMethods
   where
-    ty = mkHsWildCardBndrs $ mkHsImplicitBndrs $ mkQualType "Implements" `mkAppTy` rdrNameToType template `mkAppTy` rdrNameToType interface
-    name = noLoc $ mkRdrUnqual $ mkVarOcc ("_implements_" ++ rdrNameToString template ++ rdrNameToString interface)
-
-mkInterfaceInstance :: Located RdrName -> Located RdrName -> LHsLocalBinds GhcPs -> ImplementsDeclBlock -> LHsDecl GhcPs
-mkInterfaceInstance templateName conName sharedBinds ImplementsDeclBlock{..} =
-    instDecl $ classInstDecl
-      (mkHsAppTy (rdrNameToType $ mkInterfaceClass implementsInterface) (mkTemplateType $ fmap (occNameString . rdrNameOcc) templateName))
-      (listToBag $
-         [ mkTemplateClassMethod (occNameString $ rdrNameOcc $ unLoc name) args expr (allLetBindings True args sharedBinds)
-         | L _ (ImplementsFunction (name, expr)) <- implementsDefs
-         , let args = [asPatRecWild "this" conName]
-         ] ++ interfaceMethods implementsInterface
-      )
-
-mkInterfaceClass :: Located RdrName -> Located RdrName
-mkInterfaceClass name = L (getLoc name) $ Unqual $ mkTcOcc $ "Is" ++ occNameString (occName $ unLoc name)
+    implementsInstance =
+      instDecl
+        (classInstDecl
+          implementsConstraint
+          (listToBag interfaceMethods)
+        )
+    implementsInterfaceMethods =
+      [ let fullMethodName =
```

Comment:
maybe worth trying to factor this out into a separate method

Reply:
just to be sure, do you mean `implementsInterfaceMethods` or `fullMethodName`?

Reply:
I was thinking of factoring out the part of the list comprehension that operates on each individual item into a function and then turning the list comprehension into a one-liner/call to `map` that uses that function.